### ISSUE TITLE
scaletempo=speed=pitch broken?

### ISSUE THREAD

**Issue by kokoko3k** [mpv-player/mpv#1653]
(1)
```
mpv videofile -af-add=scaletempo=speed=pitch --speed=2
```
expected: high pitch sound and speed doubled.
but got: slow audio, fast video, desync. pitch is preserved.

(2)
```
mpv videofile -af-add=scaletempo=speed=both --speed=2
```
expected: video and audio speedup with higher pitch.
but got: video and audio speedup with the same pitch.

Am i missing something or it is a bug?

EDIT:
ok, i managed to make (2) work by setting --audio-pitch-correction=no.
but (1) is still broken.

**Comment by wm4** [comment]
Well, it's known that some of these modes are not properly usable with video. How many of these are regressions from MPlayer?

**Comment by kokoko3k** [comment]
You're right, video files behaves strange in mplayer too.
The regression is evident with audio files.
```
mplayer mp3.mp3 -af scaletempo=speed=pitch -speed 2
```
will play, as expected, the music at the same tempo, but with the pitch doubled
mpv play the file at half of the speed (!) retaining the pitch 

```
mplayer Documenti/Musica/03\ -\ Killer\ Queen.ogg  -af scaletempo=speed=both -speed 2
```
will double, as expected, the tempo AND the pitch.
mpv leaves the pitch unchanged and doubles the tempo

Also,in mpv, -af scaletempo=both produces the exact same result as -af scaletempo=both as long as you don't specify --audio-pitch-correction=no.
That way, -af scaletempo=speed=both works as expected in mpv too.

Long story short, if i'm right, it is actually not possible, in mpv, to change the pitch while retaining the tempo.
This may be useful with audio files, when you want to alter the pitch for musical purposes, like playing the piano or the guitar into a different key or singing over a difficult song and so on.

Apart from that, my personal thinking is that the scaletempo filter should behave coherently, and in mplayer it does indeed respect what you set in its options.
In mpv this does not happen (eg: --audio-pitch-correction taints it in funny way)

**Comment by wm4** [comment]
Better now? Anyway, I can barely make sense out of this behavior.

The ``--audio-pitch-correction=no`` thing is a given, because it inserts scaletempo by default. After this change it will probably insert 2 scaletempos in some cases, and mess up hard because only 1 one them might get the speed change, so you still need it.

**Comment by kokoko3k** [comment]
I repeated the very same commands of my my previous posts, but got the very same results.
Nothing seems to be changed since my previous tests. i tried mpv git-5cddd4c

That said, i don't get why audio-pitch-correction is a good thing.
I did some benchmarks to see the cpu impact of the scaletempo filter while playing media at default speed, and it seems to be zero:
with a wav file during 1 hour, entirely cached by the kernel:
```
$ sudo mplayer &>/dev/null ; sh -c "time sudo nice -n -20 mplayer longfile.wav  -ao pcm:file=/dev/null -endpos 3600 &>/dev/null"
real    0m0.453s
user    0m0.300s
sys     0m0.150s

$ sudo mplayer &>/dev/null ; sh -c "time sudo nice -n -20 mplayer longfile.wav  -ao pcm:file=/dev/null -endpos 3600 -af scaletempo &>/dev/null"
real    0m0.452s
user    0m0.303s
sys     0m0.143s
```
Now see what happens if we change the speed:
```
koko@TravelMate ~ $ sudo mplayer &>/dev/null ; sh -c "time sudo nice -n -20 mplayer longfile.wav  -ao pcm:file=/dev/null -endpos 3600 -speed 1.1 &>/dev/null"
real    0m8.037s
user    0m7.910s
sys     0m0.120s

koko@TravelMate ~ $ sudo mplayer &>/dev/null ; sh -c "time sudo nice -n -20 mplayer longfile.wav  -ao pcm:file=/dev/null -endpos 3600 -speed 1.1 -af scaletempo=speed=tempo &>/dev/null"
real    0m27.893s
user    0m27.717s
sys     0m0.163s
```

..as you can see, scaletempo has a cpu impact only if the speed is changed, so i don't get why one cannot insert scaletempo the way he likes (preserve pitch or preserve tempo) and forget it.

**Comment by kokoko3k** [comment]
If i understood the --audio-pitch-correction logic, then i can say that i've got similar results from mpv; no scaletempo overhead on speed=1
```
$ sudo mpv &>/dev/null ; sh -c "time sudo nice -n -20 mpv longfile.wav  -ao pcm:file=/dev/null --length 3600 --audio-pitch-correction=yes &>/dev/null"
real    0m1.046s
user    0m1.647s
sys     0m0.350s

$sudo mpv &>/dev/null ; sh -c "time sudo nice -n -20 mpv longfile.wav  -ao pcm:file=/dev/null --length 3600 --af-add=scaletempo=speed=tempo,scaletempo=speed=tempo,scaletempo=speed=tempo,scaletempo=speed=tempo,scaletempo=speed=tempo &>/dev/null"

real    0m1.078s
user    0m1.663s
sys     0m0.357s



$ sudo mpv &>/dev/null ; sh -c "time sudo nice -n -20 mpv longfile.wav  -ao pcm:file=/dev/null --length 3600 --audio-pitch-correction=yes --af-add=scaletempo=speed=tempo &>/dev/null"
real    0m1.026s
user    0m1.610s
sys     0m0.353s

$ sudo mpv &>/dev/null ; sh -c "time sudo nice -n -20 mpv longfile.wav  -ao pcm:file=/dev/null --length 3600 --audio-pitch-correction=no --speed=1.1 &>/dev/null"
real    0m3.133s
user    0m4.230s
sys     0m0.480s

$ sudo mpv &>/dev/null ; sh -c "time sudo nice -n -20 mpv longfile.wav  -ao pcm:file=/dev/null --length 3600 --audio-pitch-correction=yes --speed=1.1 &>/dev/null"
real    0m28.274s
user    0m29.973s
sys     0m0.447s

```

**Comment by wm4** [comment]
>I repeated the very same commands of my my previous posts, but got the very same results.

Well, no idea then, and frankly I don't even care.

>That said, i don't get why audio-pitch-correction is a good thing.

Because users requested it, and me and many others had scaletempo in their config for a long time? Other players also apply pitch correction by default. Just because it doesn't work on your Celeron from 2003 or something it doesn't mean it's bad.

>..as you can see, scaletempo has a cpu impact only if the speed is changed, so i don't get why one cannot insert scaletempo the way he likes (alters pitch or alters tempo) and forget it.

The filter has logic to skip processing if the parameter is set to identity:
```C
  if (s->scale == 1.0) {
    return data;
  }
```

**Comment by kokoko3k** [comment]
Given that scaletempo has no impact on performances as long as speed is=1, and given that some functionalities now are missing (i tested 3 systems, not just my celeron), i just can't understand why add an option to add scaletempo, when you can add scaletempo directly.
A big "?" is floating in mi mind :D

**Comment by wm4** [comment]
>Why not just insert scaletempo by default instead of setting a new option that add scaletempo? As you said, the logic to bypass the filter is already in the filter itself; no need to duplicate it.

Because inserting the filter actually does have an effect. MPlayer also needs to reconfigure the filter chain, while mpv doesn't (which means there are stricter requirements on the format passed through by the filter even if no filtering is done). mpv doesn't because that will give us smoother transitions between different playback speeds.

**Comment by kokoko3k** [comment]
Well, ok, thank you anyway for the time you spent on this thing.
I'll try to get pitch shifting to work with ladspa filters.

I don't know if this needs to be closed or not; as i understood you are not happy about your last commit on this; but given that it doesn't change a thing on my particular case (i don't know why), for me it is ok if you revert it.

**Comment by wm4** [comment]
closing because there's nothing more to do on this one.

**Comment by wm4** [comment]
Actually, looking at the code again, what I said about the ``--audio-pitch-correction`` logic was wrong. I changed it again.

**Comment by kokoko3k** [comment]
Indeed, pitch shifting with scaletempo is working again.
Thank you wm4.
